File: packages/tools/src/lib/git.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

type GitCommandOptions = {
  cwd: string;
};

export async function getRepoRoot(cwd: string): Promise<string> {
  return await runGit(["rev-parse", "--show-toplevel"], { cwd });
}

export async function getWorktreeStatus(repoRoot: string): Promise<string> {
  return await runGit(["status", "--porcelain"], { cwd: repoRoot });
}

export async function hasTag(repoRoot: string, tag: string): Promise<boolean> {
  const stdout = await runGit(["tag", "--list", tag], {
    cwd: repoRoot,
  });

  return stdout === tag;
}

export async function stageFiles(
  repoRoot: string,
  filePaths: readonly string[],
): Promise<void> {
  await runGit(["add", ...filePaths], { cwd: repoRoot });
}

export async function createCommit(
  repoRoot: string,
  message: string,
): Promise<void> {
  await runGit(["commit", "-m", message], { cwd: repoRoot });
}

export async function createTag(
  repoRoot: string,
  tag: string,
  message: string,
  options?: {
    sign: boolean;
  },
): Promise<void> {
  const tagMode = options?.sign === false ? "-a" : "-s";

  await runGit(["tag", tagMode, tag, "-m", message], { cwd: repoRoot });
}

async function runGit(
  args: readonly string[],
  options: GitCommandOptions,
): Promise<string> {
  try {
    const result = await execFileAsync("git", [...args], {
      cwd: options.cwd,
      encoding: "utf8",
    });

    return result.stdout.trim();
  } catch (error) {
    throw new Error(formatGitError(args, error));
  }
}

function formatGitError(args: readonly string[], error: unknown): string {
  const command = `git ${args.join(" ")}`;

  if (isRecord(error)) {
    const message = readStringProperty(error, "message");
    const stderr = readStringProperty(error, "stderr");

    if (stderr) {
      return `${command} failed: ${stderr.trim()}`;
    }

    if (message) {
      return `${command} failed: ${message}`;
    }
  }

  return `${command} failed`;
}

function readStringProperty(
  value: Record<string, unknown>,
  propertyName: "message" | "stderr",
): string | undefined {
  const propertyValue = value[propertyName];

  return typeof propertyValue === "string" ? propertyValue : undefined;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

File: packages/tools/lib/src/lib/git.dart
import 'dart:io';

import 'error.dart';

/// Runs a git subcommand and returns the raw [ProcessResult].
typedef GitRunner =
    Future<ProcessResult> Function(List<String> args, String workingDirectory);

Future<ProcessResult> _runGitProcess(
  List<String> args,
  String workingDirectory,
) {
  return Process.run(
    'git',
    args,
    workingDirectory: workingDirectory,
    runInShell: false,
  );
}

/// The default runner that spawns a real `git` process.
final GitRunner defaultGitRunner = _runGitProcess;

/// Process runner used by the git helpers. Overridable for tests.
GitRunner gitRunner = _runGitProcess;

Future<String> getRepoRoot(String cwd) {
  return _runGit(['rev-parse', '--show-toplevel'], cwd);
}

Future<String> getWorktreeStatus(String repoRoot) {
  return _runGit(['status', '--porcelain'], repoRoot);
}

Future<bool> hasTag(String repoRoot, String tag) async {
  final stdout = await _runGit(['tag', '--list', tag], repoRoot);

  return stdout == tag;
}

Future<void> stageFiles(String repoRoot, List<String> filePaths) async {
  await _runGit(['add', ...filePaths], repoRoot);
}

Future<void> createCommit(String repoRoot, String message) async {
  await _runGit(['commit', '-m', message], repoRoot);
}

Future<void> createTag(
  String repoRoot,
  String tag,
  String message, {
  bool sign = true,
}) async {
  final tagMode = sign ? '-s' : '-a';

  await _runGit(['tag', tagMode, tag, '-m', message], repoRoot);
}

Future<String> _runGit(List<String> args, String workingDirectory) async {
  final command = 'git ${args.join(' ')}';
  ProcessResult result;

  try {
    result = await gitRunner(args, workingDirectory);
  } on ProcessException catch (error) {
    final message = error.message.trim();

    throw ToolException(
      message.isEmpty ? '$command failed' : '$command failed: $message',
    );
  }

  if (result.exitCode != 0) {
    final stderrText = (result.stderr as String).trim();

    throw ToolException(
      stderrText.isEmpty ? '$command failed' : '$command failed: $stderrText',
    );
  }

  return (result.stdout as String).trim();
}
